### DISCUSSION TITLE
CVode Reactor for a Singles cell

### DISCUSSION THREAD

**Discussion by FernandoIGuerra** [AMReX-Combustion/PeleC#874] [Q&A]
Hi, everyone,
I was wondering if it is possible to call the reactor CVode or any chemical reactor for a single cell. Currently, the reactor takes a box as an input leading that all the box integrates using the same time step. That box comes from the MFIter loop (React.cpp)

I am trying to implement the following algorithm. To achieve so, I need that each cell react at is own characteristic time step that depends of transport coefficients. Thanks in advance.

![Screenshot from 2024-12-02 15-11-46](https://github.com/user-attachments/assets/1b8e5c33-5e47-4a86-82ec-068e06b179ed)

```
ReactorCvode::react(
  const amrex::Box& box,
  amrex::Array4<amrex::Real> const& rY_in,
  amrex::Array4<amrex::Real> const& rYsrc_in,
  amrex::Array4<amrex::Real> const& T_in,
  amrex::Array4<amrex::Real> const& rEner_in,
  amrex::Array4<amrex::Real> const& rEner_src_in,
  amrex::Array4<amrex::Real> const& `FC_in,`
  amrex::Array4<int> const& mask,
  amrex::Real& dt_react,
  amrex::Real& time
#ifdef AMREX_USE_GPU
  ,
  amrex::gpuStream_t stream
#endif
)
```

**Comment by baperry2** [comment]
The reactors in PelePhysics are designed to advance boxes of cells together because advancing each cell individually would lead to thread divergence and performance degradation on GPU. I don't fully follow the partial algorithm you posted, but it seems to diverge significantly from what is supported in the existing Reactor classes. You may need to define a new reactor class that derives from ReactorBase, which would give you more flexibility in how to compute the reaction rates.
